sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -3833,7 +3833,9 @@ class Dataset[T] private[sql](
    * @since 1.6.0
    */
   def persist(newLevel: StorageLevel): this.type = {
-    sparkSession.sharedState.cacheManager.cacheQuery(this, None, newLevel)
+    if (newLevel != StorageLevel.NONE) {
+      sparkSession.sharedState.cacheManager.cacheQuery(this, None, newLevel)
+    }
```

Comment:
Shouldn't this not be done in `cacheQuery` instead ?

Reply:
Yes, that is probably better. Updated the PR.